maintainers/patches

```diff
@@ -42,44 +27,99 @@ get_head() {
   git rev-parse HEAD
 }
 
+help() {
+  local root="$(get_root)"
+  cat <<EOF
+Usage: ${PROGRAM} {apply|save}
+
+  apply       Applies the patches to the Tock submodule regardless of its state.
+              As a consequence this can always be called to get to a clean state
+              but may result in data loss if there are unsaved changes.
+
+  save        Saves the Tock submodule to the patches.
+              This should only be called after apply and when all changes have
+              been added to a commit. After saving, you can run ./setup.sh to
+              return to normal state. Otherwise you can continue editing Tock
+              and calling save.
+
+Example:
+
+  1. Enter the edit state from the normal state:
+
+    ${PROGRAM} apply
+
+  2. Edit files in the Tock submodule:
+
+    cd third_party/tock
+    edit <files>
+
+  3. Create a fix commit per affected patch by repeating the following commands
+  until there are no more files to add:
+
+    git add -p
+    git commit -m'fix <patch#>'
+
+  4. Merge the fixes into their patches by moving their line below their patch
+  and changing their "edit" into "fixup":
+
+    git rebase -i ${root}
+
+  5. Save the changes:
+
+    cd ../..
+    ${PROGRAM} save
+
+  6. Either continue repeating steps 2 to 5, or return to the normal state:
+
+    ./setup.sh
+EOF
+  exit 0
+}
+
 apply() {
-  git submodule -q update third_party/tock
```

Comment:
Just double checking: is the `git checkout` below replacing the `git submodule update` here?

Reply:
Yes, the reason is that if the submodule is dirty, then the `git submodule update` fails. This seems to be ok if there is nothing to update which is why it works from a normal state (i.e. after setup.sh). But I would like `apply` to work it all cases such that it's easy to restore to an edit state. We could argue that we can always to `reset && setup && apply` but that's more costly. Actually I should probably add a reset command (in addition to apply and save) to have a faster setup.sh. I'll take a look into that.